### ISSUE TITLE
[Enhancement]: [a11y] ensure `abbr` element used for required fields is not announced to Screen Readers

### ISSUE THREAD

**Issue by frosso** [woothemes/woocommerce#48948]
### Describe the solution you'd like

We could add `aria-hidden` to the generated `abbr` element.
[It was initially discussed here as a possibility](https://github.com/woocommerce/woocommerce/pull/48371#pullrequestreview-2118964097). However, the addition of this attribute to the element also breaks E2E tests.
Changing the E2E tests implementation across multiple repositories can be quite tricky.
Often, E2E tests run across multiple versions of WooCommerce Core. We would need to add a getter fallback for every field with `*`, so it could be retriever without `*`.
But such change would require a large coordinated effort.

### Describe alternatives you've considered

An alternative, short-term-solution, could be to generate two labels - one with the `abbr[aria-hidden]` and another label without the `aria-hidden` attribute on the `abbr`.

The label without the `aria-hidden` on the `abbr` element could be rendered off-screen via CSS, and we could add `aria-hidden` on the label itself.
This way, the E2E tests should still work (because it would be targeting this off-screen label), and the screen readers would stop reading the "required" `abbr`.

### Additional context

With the addition of the `aria-required` attribute on required elements ( https://github.com/woocommerce/woocommerce/pull/48371 ), the `abbr` element used on the label of required fields is now duplicated.

**Comment by amesplant** [comment]
@frosso 
I am looking into this issue as an Accessibility expert, I have what might be a different solution to propose, but not sure how it would impact the E2E tests. 

Unfortunately, the tooltip that’s provided to sighted mouse users when they hover over an `<abbr> `element with a title attribute is not provided to sighted keyboard or touch users. Also, screen reader support for `<abbr>` with title is inconsistent.

I might suggest that instead, we do something along the lines of:
```
<label for="name" class="required" aria-label="First Name, required">First Name</label>
<input id="name" required>
```
and then using CSS to add the  * behind the label.

That way when users using a screen reader have focus on the input field, the label is announced with the additional "required" text, without relying on `aria-hidden` (which is not respected by all screen readers, ahem NVDA)

supporting documents: 
https://govtnz.github.io/web-a11y-guidance/wct/abbreviations/
https://a11y-101.com/development/required

**Comment by frosso** [comment]
@amesplant thank you so much for dropping in - I think that's a great idea, it solves the problem beautifully!
With this change, the markup would be different than before - something that might break some themes or extensions. But I don't think it'll be a major drawback!

Thanks again! 💯

**Comment by amesplant** [comment]
@frosso 
Your work on this looks amazing by the way, and I am trying to wrap my head around the issue in its entirety. 

It looks as though your PR: [feat: add aria-required attributes to WC form fields #48371
](https://github.com/woocommerce/woocommerce/pull/48371) is ready from the stand point of adding the `required` attributes onto each input. Is that correct? 

Does it make sense for me to create a branch off of yours and implement the solution above?

**Comment by amesplant** [comment]
@frosso I decided to create my own branch and only focus on the accessibility of the 'required' being announced by screen readers. I would love for you to take a look a the PR I created (as a draft) and let me know if you want me to continue heading in this direction.

I am testing to see if a field is required, and if it is ..
1. Adding the 'required' attribute to the field (which in of itself will announce to screen readers that the field is required)
2. Adding the asterisk, using a bit of html and css (which are empty and thus will not communicate anything to a screen reader)
3. For those fields that don't announce 'required', will attach some 'hidden' text to make sure that "required" does get announced.

I am off now until Tuesday, in which I will pick this up on my return.

**Comment by vikrampm1** [comment]
@frosso this is currently blocked and awaiting your review of the PR to proceed further.

**Comment by frosso** [comment]
@amesplant my apologies about the delay here, I completely missed your last ping!

> Does it make sense for me to create a branch off of yours and implement the solution above?
> I decided to create my own branch and only focus on the accessibility of the 'required' being announced by screen readers. I would love for you to take a look a the PR I created (as a draft) and let me know if you want me to continue heading in this direction

I think that's the right direction 👍 

I checked out your PR, and I think it's a good starting point, although it appears to be heading in a different way than I expected [from this initial comment](https://github.com/woocommerce/woocommerce/issues/48948#issuecomment-2234397318). I am not seeing the changes to the `aria-label` attribute.

A few things, before you open the PR for review:
- Please make sure the code adheres to the WC Core styleguide (usage of camelCase vs snake_case)
- Please make sure the changes are also reflected in the JS - e.g.: like in this file: `plugins/woocommerce/client/legacy/js/frontend/address-i18n.js`. This is executed on shortcode checkout when the customer changes the billing country from one that requires a "state/province" value to one that doesn't (or vice-versa).
- The addition of the `required` attribute adds front-end browser validation to the form - was that an intended consequence?

Thank you for your time and patience!

**Comment by amesplant** [comment]
@frosso I opted for a 'screen reader' span instead as it is better for translators, therefore no longer requiring the `aria-label` attribute. 

Thank you for the additional notes, I will makes sure to follow those guidelines.

The addition of the `required` attribute allows the screen reader to read the input as required, which does have the consequence of front-end browser validation.

**Comment by frosso** [comment]
> The addition of the `required` attribute allows the screen reader to read the input as required, which does have the consequence of front-end browser validation.

In https://github.com/woocommerce/woocommerce/pull/48371 we used the `aria-required` attribute approach as an alternative to the `required` attribute on inputs, because of the breaking behavior with front-end browser validation.

Do you think the `required` attribute would still be needed?

**Comment by amesplant** [comment]
@frosso looking into it now .. let me review and get back you since it's been a few days 😬

**Comment by amesplant** [comment]
@frosso I updated my PR with #49696, resolving any conflicts. The `aria-required` attribute is perfect, and so really only needed to add the asterisk as empty spans using css. This ONLY is relevant to form using the `wc-template-functions.php` at the moment.

**Comment by amesplant** [comment]
Should we also be addressing the `templates/myaccount/form-edit-account.php`?

**Comment by frosso** [comment]
Wonderful, thank you @amesplant !

> Should we also be addressing the templates/myaccount/form-edit-account.php?

That's a good point, and it's something I didn't consider before adding `aria-required` to checkout fields.
It would be excellent to have a consistent experience across the whole application.
I'll be updating the other templates with the `aria-required` attribute 👍

**Comment by amesplant** [comment]
@frosso happy to take on the `aria-required` and the asterisk in `templates/myaccount/form-edit-account.php` in this PR if you'd like 🤓

**Comment by amesplant** [comment]
@frosso added ^^^^

**Comment by frosso** [comment]
Thanks @amesplant , that's great! I think it's looking pretty good, feel free to open the PR for review once it's ready. I can take a first pass of review if you'd like, but somebody else from the WooCommerce Core contributor team will need to approve & merge the PR 👍

**Comment by amesplant** [comment]
Thanks @frosso 
@dkotter this one is ready for code review before we pass it back to Woo 😄
